### PR TITLE
Fix failed starts

### PR DESCRIPTION
**Pull request by FFroehlich**
Includes #315

### CODE REVIEW

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_history.py  29
- pypesto/objective/amici_calculator.py  6
         

Complexity decreasing per file
==============================
+ pypesto/objective/history.py  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_history.py  31
- pypesto/objective/amici_calculator.py  6
         

Complexity decreasing per file
==============================
+ pypesto/objective/history.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_history.py  29
- pypesto/objective/amici_calculator.py  6
         

Complexity decreasing per file
==============================
+ pypesto/objective/history.py  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_history.py  5
- pypesto/objective/amici_calculator.py  6
         

Complexity decreasing per file
==============================
+ pypesto/objective/history.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_history.py  5
- pypesto/objective/amici_calculator.py  6
         

Complexity decreasing per file
==============================
+ pypesto/objective/history.py  -6
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/objective/amici_calculator.py  6
- test/test_history.py  5
         

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
+ pypesto/objective/history.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

File: pypesto/optimize/optimizer.py [hunk 1]
```diff
@@ -68,9 +81,11 @@ def time_decorator(minimize):
     the wall-clock time.
     """
 
-    def wrapped_minimize(self, problem, x0, id, history_options=None):
+    def wrapped_minimize(self, problem, x0, id, allow_failed_starts,
```
Comment (yannikschaelte) [inline pypesto/optimize/optimizer.py:84]: These wrappers are getting a little complicated, as we have not at least two arguments that are not needed for the "minimization" task, just in the wrappers. I suggest reformulating this as functions of Optimizer at some point, but fine for the moment.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/objective/history.py  3
         

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
+ pypesto/objective/history.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/objective/history.py  6
         

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/objective/history.py  6
         

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/objective/history.py  6
         

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/objective/history.py  6
         

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by FFroehlich** [comment]
@yannikschaelte did substantial changes after review, want to have another look or fine to merge?

File: pypesto/objective/amici_calculator.py [hunk 2]
```diff
@@ -23,6 +25,7 @@ class AmiciCalculator:
     Class to perform the actual call to AMICI and obtain requested objective
     function values.
     """
+    _known_least_squares_safe = False
```
Comment (yannikschaelte) [inline pypesto/objective/amici_calculator.py:28]: this way, it becomes a class attribute, not an object attribute ... might be better to define inside init

File: pypesto/objective/history.py [hunk 3]
```diff
@@ -882,89 +884,100 @@ def update(self,
...
+        sres = result.get(SRES, None)
+        if fval is not None and fval < self.fval_min:
+            self.fval_min = fval
+            self.x_min = x
+            self.grad_min = grad
+            self.hess_min = hess
+            self.res_min = res
+            self.sres_min = sres
+
+        x_same = False
```
Comment (yannikschaelte) [inline pypesto/objective/history.py:919]: can we just have `x_same = np.all(self.x_min == x)`` minimum (?) computational effort, better readable

Comment (yannikschaelte) [reply pypesto/objective/history.py:919]: and add a comment that the following is due to gradients often being evaluated only after a good point has been found, in subsequent iterations?

File: pypesto/objective/util.py [hunk 4]
```diff
@@ -13,7 +13,7 @@ def res_to_chi2(res: np.ndarray):
     """
     if res is None:
         return None
-    return 0.5 * np.power(res, 2).sum()
+    return np.dot(res, res)
```
Comment (yannikschaelte) [inline pypesto/objective/util.py:16]: why remove the 0.5? I think I've seen both implementations in use.

Comment (FFroehlich) [reply pypesto/objective/util.py:16]: We now use the chi2 value provided by AMICI, which doesn't have the 0.5 factor. I would actually prefer to have the 0.5 factor here but find it more important to be consistent with what amici returns

Comment (FFroehlich) [reply pypesto/objective/util.py:16]: will add comment

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/engine/task.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5495896&bid=18009890)

File: pypesto/objective/util.py (hunk 4 above)
Comment (yannikschaelte) [reply pypesto/objective/util.py:16]: Should be fine I guess (not aware of having used this factor anywhere explicitly, except maybe in a test. Please also update the docstring above.

Comment (FFroehlich) [reply pypesto/objective/util.py:16]: yep, thats what I did :)

Comment (yannikschaelte) [reply pypesto/objective/util.py:16]: ok I was too late. all done already :)

**Pull request merged**